File: elixir/resistor-color-trio/lib/resistor_color_trio.ex
defmodule ResistorColorTrio do
  @doc """
  Calculate the resistance value in ohm or kiloohm from resistor colors
  """

  @kilo 1_000
  @mega 1_000_000
  @giga 1_000_000_000

  @codes %{
    black: 0,
    brown: 1,
    red: 2,
    orange: 3,
    yellow: 4,
    green: 5,
    blue: 6,
    violet: 7,
    grey: 8,
    white: 9
  }

  @spec label(colors :: [atom]) :: {number, :ohms | :kiloohms}
  def label(colors) do
    [first, second, third | _] = colors

    ((@codes[first] * 10 + @codes[second]) * Integer.pow(10, @codes[third]))
    |> format_resistance()
  end

  defp format_resistance(resistance) when resistance > @giga,
    do: {div(resistance, @giga), :gigaohms}

  defp format_resistance(resistance) when resistance > @mega,
    do: {div(resistance, @mega), :megaohms}

  defp format_resistance(resistance) when resistance > @kilo,
    do: {div(resistance, @kilo), :kiloohms}

  defp format_resistance(resistance), do: {resistance, :ohms}
end

File: python/resistor-color-trio/resistor_color_trio.py
GIGA = 1_000_000_000
MEGA = 1_000_000
KILO = 1_000

encoding = {
    "black": 0, 
    "brown": 1, 
    "red": 2, 
    "orange": 3, 
    "yellow": 4, 
    "green": 5, 
    "blue": 6, 
    "violet": 7, 
    "grey": 8, 
    "white": 9
}

def format_output(value):
    
    if (value >= GIGA) :
            return f'{int(value / GIGA)} gigaohms'

    if (value >= MEGA) :
            return f'{int(value / MEGA)} megaohms'

    if (value >= KILO) :
            return f'{int(value / KILO)} kiloohms'
        
    return f'{value} ohms'

def label(colors):
     resistance = encoding[colors[0]] * 10 + encoding[colors[1]]
     resistance *= pow(10, encoding[colors[2]])

     return format_output(resistance)
